src/skmultiflow/anomaly_detection/half_space_trees.py

```diff
@@ -0,0 +1,548 @@
+            The maximum depth of the trees in the ensemble.
+            'maxDepth' in the original paper.
+
+        size_limit: int, optional (default=50)
+            The minimum mass required in a node (as a fraction of the window size) to calculate the anomaly score.
+            'sizeLimit' in the original paper.
+        """
+        super().__init__()
+        self.max_depth = max_depth
+        self.dimensions = dimensions
+        self.root = HalfSpaceTreeNode()
+        self.size_limit = size_limit
+        self.is_learning_phase_on = True
+
+    def predict_proba(self, X, max_score):
+        """ Predicts probabilities of all label of the X instance(s)
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (n_samples, n_features)
+            Samples for which we want to predict the labels.
+
+        max_score: float
+            The maximum score of an instance could have in the tree.
+
+        Returns
+        -------
+        numpy.array
+            Predicted the probabilities of all the labels for all instances in X.
+
+        """
+        r, _ = get_dimensions(X)
+        predictions = []
+        for i in range(r):
+            votes = self.get_votes_for_instance(X[i], max_score).copy()
+            y_proba = np.zeros(int(max(votes.keys())) + 1)
+            for key, value in votes.items():
+                y_proba[int(key)] = value
+                predictions.append(y_proba)
+        return np.array(predictions)
+
+    def get_votes_for_instance(self, X, max_score):
+        """ Get class votes for a single instance.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of length equal to the number of features.
+            Instance attributes.
+
+        max_score: float
+            The maximum score of an instance could have in the tree.
+        Returns
+        -------
+        dict (class_value, weight)
+
+        """
+        if self.root is not None and self.is_learning_phase_on is not True:
+            score = self.anomaly_score(X)
+            anomaly_score = 1 - score / max_score
+            return {0: 1 - anomaly_score, 1: anomaly_score}
+        else:
+            return {0: 0.5, 1: 0.5}
+
+    def anomaly_score(self, X):
+        """
+        Calculates anomaly score of a given instance X.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (1, n_features)
+            Instance attributes.
+
+        Returns
+        -------
+
+        """
+        return self._anomaly_score(self.root, X)
+
+    def _anomaly_score(self, node, X):
+        """
+        Private function to calculate anomaly score.
+
+        Parameters
+        ----------
+        node: HalfSpaceTreeNode
+            Current node
+
+        X: numpy.ndarray of shape (1, n_features)
+            Instance attributes.
+
+        Returns
+        -------
+        anomaly_score: float
+            Anomaly score of the instance X.
+
+        """
+        anomaly_score = 0.5
```

Comment:
Should this be `anomaly_score = 0.0`?

Reply:
Even tough we define a threshold to decide whether the sample is anomalous or not, I believe it is more humanly to say if the score is closer to zero it's not an anomaly and if it is closer to one it is an anomaly. I would say that in this logic, the middle point should be quantified as the default value.

Reply:
Looking closer into this, it seems that this value is actually never used. This would only happen if `node is None` which does not seem possible (given the structure of the code.)

Reply:
Yes it is true, since the node never will be `None`. I can remove it then?